File: session.cpp
#include "session.hpp"
#include <chrono>
#include <httplib.h>

#include "lib/podmanClient.hpp"
#include <iostream>
#include <sstream>
#include "lib/lib/socket.hpp"

PodmanClient podmanClient(std::getenv("PODMAN_SOCKET") == nullptr ? "http://localhost:8888" : std::getenv("PODMAN_SOCKET"));

std::string getImageTag(int session, int id) {
    return std::to_string(std::chrono::system_clock::now().time_since_epoch().count()) + "-" + std::to_string(session) + "-" + std::to_string(id);
}

std::string getContainerName(int session, int id, int image) {
    return std::to_string(std::chrono::system_clock::now().time_since_epoch().count()) + "_" + std::to_string(session) + "_" + std::to_string(image) + "_" + std::to_string(id);
}

int findFreePort(int min, int max = 65535) {
    for (int port = min; port <= std::min(max, 65535); port++) {
        httplib::Server svr;

        // Try to bind to the port
        if (svr.bind_to_port("0.0.0.0", port)) {
            // Success - port is available
            svr.stop();
            return port;
        }

        // Port in use, try next one
        svr.stop();
    }

    throw std::runtime_error("No available port found >= " + std::to_string(min));
}

int Session::sessionsCount = 0;

Session::Session(const std::map<std::string, std::string>& networks, const std::shared_ptr<Socket::Connection*>& connection, const int id): id(id), connection(connection), networks(networks) {}

void Session::onData(std::string data) {
    std::cout << "Received: " << data << '\n' << std::endl;
    std::istringstream stream(data);
    std::string type; stream >> type;
    if (type == "BUILD") {
        int imageId; stream >> imageId;
        std::string context, dockerfile;
        std::getline(stream, context);
        std::getline(stream, context);
        std::getline(stream, dockerfile);
        build(imageId, context, dockerfile);
    }
    else if (type == "RUN") {
        int id, imageId; stream >> id >> imageId;
        std::string stdout = "normal", stderr = "onEnd", subtype;
        std::vector<std::pair<std::string, std::string>> volumes;
        std::map<std::string, std::string> env;
        std::vector<std::string> networks;
        std::string initStdin;
        while (stream >> subtype) {
            if (subtype == "STDOUT") stream >> stdout;
            else if (subtype == "STDERR") stream >> stderr;
            else if (subtype == "VOLUME") {
                std::string from, to;
                std::getline(stream, from);
                std::getline(stream, to);
                from.erase(0, 1);
                volumes.emplace_back(from, to);
            }
            else if (subtype == "ENV") {
                std::string key, value;
                stream >> key;
                std::getline(stream, value);
                value.erase(0, 1);
                env.emplace(key, value);
            }
            else if (subtype == "NETWORK") {
                std::string network;
                std::getline(stream, network);
                network.erase(0, 1);
                networks.push_back(network);
            }
            else if (subtype == "WRITE") {
                std::string tmp; std::getline(stream, tmp);
                while (std::getline(stream, tmp)) initStdin += tmp;
            }
            else {
                std::cout << "Error: " << subtype << '\n' << std::endl;
            }
        }
        run(id, imageId, stdout, stderr, networks, volumes, env, initStdin);
    }
    else if (type == "RESTART") {
        int id; stream >> id;
        restart(id);
    }
    else if (type == "STOP") {
        int id; stream >> id;
        stop(id);
    }
    else if (type == "WRITE") {
        int id; stream >> id;
        std::string buffer, tmp;
        while (std::getline(stream, tmp)) buffer += tmp;
        write(id, buffer);
    }
    else if (type == "HOST") {
        int id; stream >> id;
        getHost(id);
    }
    else if (type == "VERDICT") {
        std::string verdict, sub, subtask, data_; stream >> verdict >> sub;
        if (sub == "SUB") {
            std::getline(stream, subtask);
            stream >> sub;
        }
        if (sub == "DATA") {
            std::string tmp; std::getline(stream, tmp);
            while (std::getline(stream, tmp)) data_ += tmp;
        }

    }
    else {
        std::cout << "Unknown type: " << type << std::endl;
    }
}

void Session::build(int image, const std::string& context, const std::string& dockerfilePath) {
    auto tag = getImageTag(id, image);
    images[image] = tag;
    revImages[tag] = image;
    podmanClient.build(tag, context, dockerfilePath);
}

std::function<void(const std::string&)> stdoutCallback(int id, const std::string& stdout, std::shared_ptr<Socket::Connection*> connection) {
    return [&connection, &id](const std::string& chunk) {
        (*connection)->write("STDOUT " + std::to_string(id) + '\n' + chunk);
    };
}

void Session::run(int id, int image, const std::string& stdout, const std::string& stderr, std::vector<std::string> networks, const std::vector<std::pair<std::string, std::string>>& volumes, const std::map<std::string, std::string>& env, const std::string& initStdin) {
    for (auto& network : networks) network = this->networks[network];
    auto containerId = podmanClient.run(images[image], {}, {}, env, volumes, networks, initStdin);
    containers[id] = containerId;
    revContainers[containerId] = id;
    // if (stdout != "none" || stderr != "none") podmanClient.attach(containerId);
    // if (stdout != "none") podmanClient.onStdout(containerId, stdoutCallback(id, stdout, connection));
    // if (stderr != "none") podmanClient.onStderr(containerId, stdoutCallback(id, stderr, connection));
}

void Session::restart(int id) {
    podmanClient.restart(containers[id]);
}

void Session::stop(int id) {
    podmanClient.stop(containers[id]);
}

void Session::write(int id, const std::string& chunk) {
    podmanClient.write(containers[id], chunk);
}

void Session::getHost(int id) {
    (*connection)->write("HOST " + podmanClient.getName(containers[id]));
}

void Session::verdict(int id, const std::string& sub, const std::string& data) {

}

File: session.ts
import path from 'path';
import { Connection } from './socket';
import Task from './task';
import {podmanClient, client} from "./main";

let sessionsCount = 0;

function getImageTag(session: number, id: number): string {
    return Date.now() + '-' + session + '-' + id;
}

function getContainerName(session: number, id: number, image: number): string {
    return Date.now() + '_' + session + '_' + image + '_' + id;
}

interface Volume {
    host: string;
    container: string;
}

interface Env {
    [key: string]: string;
}

class Session {
    private id: number;
    private images: Map<number, string>;
    private containers: Map<number, string>;
    private revImages: Map<string, number>;
    private revContainers: Map<string, number>;
    private connection: Connection;
    private volumePath: string;
    networks: Record<string, string>;
    private task: Task;

    constructor(
        networks: Record<string, string>,
        volumePath: string,
        connection: Connection,
        task: Task,
        id: number = sessionsCount++
    ) {
        this.id = id;
        this.images = new Map();
        this.containers = new Map();
        this.revImages = new Map();
        this.revContainers = new Map();
        this.connection = connection;
        this.volumePath = volumePath;
        this.networks = networks;
        this.task = task;
    }

    async onData(data: string): Promise<void> {
        if (!data) return;
        console.log('Received: ' + data);
        const lines: Array<string> = data.split('\n');
        if (lines.length === 0) return;
        const firstLine: Array<string> = (lines[0] || '').split(' ');
        if (firstLine.length < 1) return;
        const type = firstLine[0];
        let lineIdx = 1;

        if (type === 'BUILD') {
            if (firstLine.length < 2 || firstLine[1] === undefined) return;
            const imageId = parseInt(firstLine[1]);
            if (isNaN(imageId)) return;
            const context = lines[lineIdx++] || '';
            const dockerfile = lines[lineIdx++] || '';
            await this.build(imageId, context, dockerfile);
        } else if (type === 'RUN') {
            if (firstLine.length < 3 || firstLine[1] === undefined || firstLine[2] === undefined) return;
            const id = parseInt(firstLine[1]);
            if (isNaN(id)) return;
            const imageId = parseInt(firstLine[2]);
            if (isNaN(imageId)) return;
            let stdout: string = 'normal';
            let stderr: string = 'onEnd';
            const volumes: Volume[] = [];
            const env: Env = {};
            const networks: string[] = [];
            let initStdin: string = '';
            let subIdx = 3;

            while (subIdx < firstLine.length) {
                const subtype = firstLine[subIdx++];
                if (subtype === 'STDOUT') {
                    if (subIdx >= firstLine.length || firstLine[subIdx] === undefined) return;
                    stdout = firstLine[subIdx++] ?? "";
                } else if (subtype === 'STDERR') {
                    if (subIdx >= firstLine.length || firstLine[subIdx] === undefined) return;
                    stderr = firstLine[subIdx++] ?? "";
                } else if (subtype === 'VOLUME') {
                    if (lineIdx >= lines.length) return;
                    const from = (lines[lineIdx++] || '').slice(1);
                    if (lineIdx >= lines.length) return;
                    const to = (lines[lineIdx++] || '').slice(1);
                    volumes.push({ host: from, container: to });
                } else if (subtype === 'ENV') {
                    if (subIdx >= firstLine.length || firstLine[subIdx] === undefined) return;
                    const key = firstLine[subIdx++];
                    if (lineIdx >= lines.length) return;
                    const value = (lines[lineIdx++] || '').slice(1);
                    if (key !== undefined) env[key] = value;
                } else if (subtype === 'NETWORK') {
                    if (lineIdx >= lines.length) return;
                    const network = (lines[lineIdx++] || '').slice(1);
                    networks.push(network);
                } else if (subtype === 'WRITE') {
                    initStdin = lines.slice(lineIdx).join('\n');
                    break;
                } else {
                    console.log('Unknown subtype: ' + subtype);
                }
            }
            await this.run(id, imageId, stdout, stderr, networks, volumes, env, initStdin);
        } else if (type === 'RESTART') {
            if (firstLine.length < 2 || firstLine[1] === undefined) return;
            const id = parseInt(firstLine[1]);
            if (isNaN(id)) return;
            await this.restart(id);
        } else if (type === 'STOP') {
            if (firstLine.length < 2 || firstLine[1] === undefined) return;
            const id = parseInt(firstLine[1]);
            if (isNaN(id)) return;
            await this.stop(id);
        } else if (type === 'WRITE') {
            if (firstLine.length < 2 || firstLine[1] === undefined) return;
            const id = parseInt(firstLine[1]);
            if (isNaN(id)) return;
            const buffer = lines.slice(1).join('\n');
            await this.write(id, buffer);
        } else if (type === 'HOST') {
            if (firstLine.length < 2 || firstLine[1] === undefined) return;
            const id = parseInt(firstLine[1]);
            if (isNaN(id)) return;
            await this.getHost(id);
        } else if (type === 'VERDICT') {
            if (firstLine.length < 2 || firstLine[1] === undefined) return;
            const verdict = firstLine[1];
            let subtask: string | null = null;
            let subIdx = 2;
            if (subIdx < firstLine.length && firstLine[subIdx] === 'SUB') {
                subIdx += 1;
                if (subIdx >= firstLine.length || firstLine[subIdx] === undefined) return;
                subtask = firstLine[subIdx] ?? "";
                subIdx += 1;
            }
            let data_ = '';
            if (subIdx < firstLine.length && firstLine[subIdx] === 'DATA') {
                data_ = lines.slice(1).join('\n');
            }
            this.verdict(verdict, subtask, data_);
        } else {
            console.log('Unknown type: ' + type);
        }
    }

    async build(image: number, context: string, dockerfilePath: string): Promise<void> {
        const tag = getImageTag(this.id, image);
        this.images.set(image, tag);
        this.revImages.set(tag, image);
        await podmanClient.build(tag, context, dockerfilePath);
    }

    async run(
        id: number,
        image: number,
        stdoutMode: string,
        stderrMode: string,
        networks: string[],
        volumes: Volume[],
        env: Env,
        initStdin: string
    ): Promise<void> {
        const mappedNetworks = networks.map((net) => this.networks[net] || net);
        const mappedVolumes = volumes.map(({ host, container }) => ({
            host: path.join(this.volumePath, host),
            container,
        }));
        const containerId = await podmanClient.create(this.images.get(image)!, [], [], env, mappedVolumes, mappedNetworks);
        this.containers.set(id, containerId);
        this.revContainers.set(containerId, id);
        await podmanClient.start(containerId, initStdin);

        if (stdoutMode !== 'none' || stderrMode !== 'none') {
            let stdoutCollected = '';
            let stderrCollected = '';
            const stdoutCb = (chunk: string) => {
                if (stdoutMode === 'normal') this.connection.write('STDOUT ' + id + '\n' + chunk);
                else if (stdoutMode === 'onEnd') stdoutCollected += chunk;
            };
            const stderrCb = (chunk: string) => {
                if (stderrMode === 'normal') this.connection.write('STDERR ' + id + '\n' + chunk);
                else if (stderrMode === 'onEnd') stderrCollected += chunk;
            };
            const stream = await podmanClient.attachOutputs(containerId, stdoutCb, stderrCb);
            stream.on('end', () => {
                if (stdoutMode === 'onEnd' && stdoutCollected) this.connection.write('STDOUT ' + id + '\n' + stdoutCollected);
                if (stderrMode === 'onEnd' && stderrCollected) this.connection.write('STDERR ' + id + '\n' + stderrCollected);
            });
        }
    }

    async restart(id: number): Promise<void> {
        await podmanClient.restart(this.containers.get(id)!);
    }

    async stop(id: number): Promise<void> {
        await podmanClient.stop(this.containers.get(id)!);
    }

    async write(id: number, chunk: string): Promise<void> {
        await podmanClient.write(this.containers.get(id)!, chunk);
    }

    async getHost(id: number): Promise<void> {
        const name = await podmanClient.getName(this.containers.get(id)!);
        this.connection.write('HOST ' + name);
    }

    verdict(verdict: string, subtask: string | null, data: string): void {
        if (subtask) {
            client.sendSubtaskVerdict(this.task.getId(), subtask, verdict, data);
        } else {
            client.sendFullVerdict(this.task.getId(), verdict, data);
        }
    }
}

export default Session;